src/core/tests/type_prop/segment_max.cpp

```diff
@@ -0,0 +1,310 @@
+    op->validate_and_infer_types();
+
+    EXPECT_EQ(op->get_output_element_type(0), element::f32);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape::dynamic()));
+    EXPECT_EQ(op->get_fill_mode(), op::FillMode::ZERO);
+}
+
+TEST_F(TypePropSegmentMaxTest, segment_ids_dynamic) {
+    const auto data = std::make_shared<Parameter>(element::f32, PartialShape{3, 12, 81});
+    const auto segment_ids = std::make_shared<Parameter>(element::i64, PartialShape::dynamic());
+    const auto num_segments = std::make_shared<Parameter>(element::i64, PartialShape{});
+
+    const auto op = make_op(data, segment_ids, num_segments, op::FillMode::ZERO);
+    op->validate_and_infer_types();
+
+    EXPECT_EQ(op->get_output_element_type(0), element::f32);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape{Dimension::dynamic(), 12, 81}));
+    EXPECT_EQ(op->get_fill_mode(), op::FillMode::ZERO);
+}
+
+TEST_F(TypePropSegmentMaxTest, num_segments_dynamic) {
+    const auto data = std::make_shared<Parameter>(element::f32, PartialShape{3, 12, 81});
+    const auto segment_ids = std::make_shared<Parameter>(element::i64, PartialShape{3});
+    const auto num_segments = std::make_shared<Parameter>(element::i64, PartialShape::dynamic());
+
+    const auto op = make_op(data, segment_ids, num_segments, op::FillMode::ZERO);
+    op->validate_and_infer_types();
+
+    EXPECT_EQ(op->get_output_element_type(0), element::f32);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape{Dimension::dynamic(), 12, 81}));
+    EXPECT_EQ(op->get_fill_mode(), op::FillMode::ZERO);
+}
+
+TEST_F(TypePropSegmentMaxTest, dynamic_dimensions) {
+    const auto data = std::make_shared<Parameter>(element::f32, PartialShape{3, Dimension::dynamic(), 81});
+    const auto segment_ids = std::make_shared<Parameter>(element::i64, PartialShape{Dimension::dynamic()});
+    const auto num_segments = std::make_shared<Parameter>(element::i64, PartialShape{});
+
+    const auto op = make_op(data, segment_ids, num_segments, op::FillMode::ZERO);
+    op->validate_and_infer_types();
+
+    EXPECT_EQ(op->get_output_element_type(0), element::f32);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape{Dimension::dynamic(), Dimension::dynamic(), 81}));
+    EXPECT_EQ(op->get_fill_mode(), op::FillMode::ZERO);
+}
+
+TEST_F(TypePropSegmentMaxTest, dynamic_num_segments_const_segment_ids) {
+    const auto data = std::make_shared<Parameter>(element::f32, PartialShape{3, Dimension::dynamic(), 81});
+    const auto segment_ids = std::make_shared<Constant>(element::i32, Shape{3}, std::vector<int32_t>{0, 1, 2});
+    const auto num_segments = std::make_shared<Parameter>(element::i32, PartialShape{});
+
+    const auto op = make_op(data, segment_ids, num_segments, op::FillMode::ZERO);
+    op->validate_and_infer_types();
+
+    EXPECT_EQ(op->get_output_element_type(0), element::f32);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape{3, Dimension::dynamic(), 81}));
```

Comment:
While `num_segments` input is a Parameter, why the `num_segments` dim in the output is static and equal `3`? The spec says it can be lower than (max_idx + 1), so I would expect dynamic dim (that can has defined upper bound to like `Dimension::dynamic(-1, 3)`.

Reply:
Fixed, now `num_segments` (when used) always overrides `segment_ids`.